### PR TITLE
enh(resources): return notes in the externalLinks

### PR DESCRIPTION
**Pull request by adr-mo**
## Description

This PR intends to retrieve the "notes" field out of the extended informations from the resources.

When calling the `/monitoring/resources `GET ENDPOINT the API should return the notes field when defined.

Example of return

```
{
    "duration": "2w 12h",
    "last_check": "1m 27s",
    "short_type": "h",
    "id": 17,
    "type": "host",
    "name": "Central",
    "alias": "Central",
    "fqdn": "localhost",
    "icon": {},
    "parent": null,
    "status": {},
    "flapping": false,
    "in_downtime": false,
    "acknowledged": false,
    "active_checks": true,
    "passive_checks": false,
    "links": {
        "uris": {},
        "endpoints": {},
        "externals": {
            "action_url": "",
            "notes": {
                "notes": "Procedure to turn back on the host",
                "notes_url": "http://mediawikit/Central/"
            }
        }
    },
    "severity": null,
    "chart_url": null,
    "last_status_change": "2021-02-08T10:29:34+01:00",
    "tries": "1/3 (H)",
    "information": "OK - localhost rta 0,132ms lost 0%",
    "performance_data": "rta=0,132ms;3000,000;5000,000;0; pl=0%;80;100;0;100 rtmax=0,132ms;;;; rtmin=0,132ms;;;; "
}
```

/!\ This PR will only be merged when @bdauria will have adapted / implemented the front-end part

**Fixes** # (issue)

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 19.10.x
- [ ] 20.04.x
- [x] 20.10.x
- [x] 21.04.x (master)

<h2> How this pull request can be tested ? </h2>

- Call the API
- Use FE to confirm that the notes is correctly handled

## Checklist

- [x] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: src/Centreon/Domain/Monitoring/Notes.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    /**
+     * @var string|null
+     */
+    private $notesUrl;
+
+    /**
+     * Get notes
+     *
+     * @return string
```
Comment (jeremyjaouen) [inline src/Centreon/Domain/Monitoring/Notes.php:46]:
```suggestion
     * @return string|null
```

File: src/Centreon/Domain/Monitoring/Notes.php [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    public function setNotes(?string $notes): self
+    {
+        $this->notes = $notes;
+        return $this;
+    }
+
+    /**
+     * Get notes url
+     *
+     * @return string
```
Comment (jeremyjaouen) [inline src/Centreon/Domain/Monitoring/Notes.php:68]:
```suggestion
     * @return string|null
```

File: src/Centreon/Domain/Monitoring/Notes.php [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+     */
+    public function getNotesUrl(): ?string
+    {
+        return $this->notesUrl;
+    }
+
+    /**
+     * Set notes url
+     *
+     * @param string $notesUrl
```
Comment (jeremyjaouen) [inline src/Centreon/Domain/Monitoring/Notes.php:78]:
```suggestion
     * @param string|null $notesUrl
```

File: src/Centreon/Domain/Monitoring/Notes.php [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+     */
+    public function getNotes(): ?string
+    {
+        return $this->notes;
+    }
+
+    /**
+     * Set notes
+     *
+     * @param string $notes
```
Comment (jeremyjaouen) [inline src/Centreon/Domain/Monitoring/Notes.php:56]:
```suggestion
     * @param string|null $notes
```

**Review by jeremyjaouen — CHANGES_REQUESTED** [review]

File: src/Centreon/Infrastructure/Monitoring/ResourceRepositoryRDB.php [hunk 5]
```diff
@@ -844,11 +847,15 @@ protected function parseResource(array $data): ResourceEntity
...
-        $resource->getLinks()->setExternals(EntityCreator::createEntityByArray(
-            ResourceExternalLinks::class,
-            $data
-        ));
+        // Setting the External links
+        $externalLinks = $resource->getLinks()->getExternals();
+        $externalLinks->setActionUrl($data['action_url']);
+        $notes = new Notes();
+        $notes->setNotesUrl($data['notes_url']);
+        $notes->setNotes($data['notes']);
```
Comment (kduret) [inline src/Centreon/Infrastructure/Monitoring/ResourceRepositoryRDB.php:855]:
```suggestion
        $notes = (new Notes())
            ->setNotesUrl($data['notes_url'])
            ->setNotes($data['notes']);
```

File: doc/API/centreon-api-v2.yaml [hunk 6]
```diff
@@ -4224,6 +4223,19 @@ components:
           type: string
           description: "Name of the group to which belongs the resource"
           example: "ESX-France|Cpus"
+    Monitoring.Notes:
+      type: object
+      properties:
+        notes_url:
```
Comment (kduret) [inline doc/API/centreon-api-v2.yaml:4229]: I think `notes` property should use following format : 
```
  notes: {
    label: 'my label',
    url: 'my url'
  }
```
cause currently `notes.notes` and `notes.notes_url` are a bit redondent

Comment (adr-mo) [reply doc/API/centreon-api-v2.yaml:4229]: alright 👍

File: src/Centreon/Infrastructure/Monitoring/ResourceRepositoryRDB.php [hunk 7]
```diff
@@ -214,7 +215,7 @@ public function findResources(ResourceFilter $filter): array
             . 'resource.parent_id, resource.parent_name, resource.parent_type, ' // parent
             . 'resource.parent_alias, resource.parent_fqdn, ' // parent
             . 'resource.parent_icon_name, resource.parent_icon_url, ' // parent icon
-            . 'resource.action_url, resource.notes_url, ' // external urls
+            . 'resource.action_url, resource.notes_url, resource.notes, ' // external urls
```
Comment (kduret) [inline src/Centreon/Infrastructure/Monitoring/ResourceRepositoryRDB.php:218]:
```suggestion
            . 'resource.action_url, resource.notes_url, resource.notes_label, ' // external urls
```

File: src/Centreon/Infrastructure/Monitoring/ResourceRepositoryRDB.php [hunk 8]
```diff
@@ -639,6 +641,7 @@ protected function prepareQueryForHostResources(StatementCollector $collector, R
             h.icon_image AS `icon_url`,
             h.action_url AS `action_url`,
             h.notes_url AS `notes_url`,
+            h.notes AS `notes`,
```
Comment (kduret) [inline src/Centreon/Infrastructure/Monitoring/ResourceRepositoryRDB.php:644]:
```suggestion
            h.notes AS `notes_label`,
```

File: src/Centreon/Infrastructure/Monitoring/ResourceRepositoryRDB.php [hunk 9]
```diff
@@ -435,6 +436,7 @@ protected function prepareQueryForServiceResources(StatementCollector $collector
             s.icon_image AS `icon_url`,
             s.action_url AS `action_url`,
             s.notes_url AS `notes_url`,
+            s.notes AS `notes`,
```
Comment (kduret) [inline src/Centreon/Infrastructure/Monitoring/ResourceRepositoryRDB.php:439]:
```suggestion
            s.notes AS `notes_label`,
```

File: src/Centreon/Infrastructure/Monitoring/ResourceRepositoryRDB.php [hunk 10]
```diff
@@ -844,11 +847,15 @@ protected function parseResource(array $data): ResourceEntity
...
-        $resource->getLinks()->setExternals(EntityCreator::createEntityByArray(
-            ResourceExternalLinks::class,
-            $data
-        ));
+        // Setting the External links
+        $externalLinks = $resource->getLinks()->getExternals();
+        $externalLinks->setActionUrl($data['action_url']);
+        $notes = new Notes();
+        $notes->setUrl($data['notes_url']);
+        $notes->setLabel($data['notes']);
```
Comment (kduret) [inline src/Centreon/Infrastructure/Monitoring/ResourceRepositoryRDB.php:855]:
```suggestion
        $notes = (new Notes())
            ->setUrl($data['notes_url'])
            ->setLabel($data['notes_label']);
```

File: src/Centreon/Domain/Monitoring/ResourceService.php [hunk 11]
```diff
@@ -271,7 +271,11 @@ private function replaceMacrosInUrlsForServiceResource(ResourceEntity $resource,
     public function replaceMacrosInExternalLinks(ResourceEntity $resource): void
     {
         $actionUrl = $resource->getLinks()->getExternals()->getActionUrl();
-        $notesUrl = $resource->getLinks()->getExternals()->getNotesUrl();
+        $notesObject = $resource->getLinks()->getExternals()->getNotes();
+        $notesUrl = null;
+        if ($notesObject !== null) {
+            $notesUrl = $notesObject->getUrl();
+        }
```
Comment (kduret) [inline src/Centreon/Domain/Monitoring/ResourceService.php:278]:
```suggestion
        $notesUrl =($notesObject !== null) ? $notesObject->getUrl() : null;
```

File: doc/API/centreon-api-v2.yaml [hunk 12]
```diff
@@ -511,6 +511,8 @@ paths:
         * status_code
         * status
         * action_url
+        * label
+        * url
```
Comment (kduret) [inline doc/API/centreon-api-v2.yaml:515]:
```suggestion
        * notes_label
        * notes_url
```

File: doc/API/centreon-api-v2.1.yaml [hunk 13]
```diff
@@ -4232,11 +4234,21 @@ components:
...
-              type: string
-              nullable: true
-              description: "URL that can be used to provide more information on the resource"
-              example: "http://mediawiki/resource/name"
+            notes:
+              $ref: '#/components/schemas/Monitoring.Notes'
+    Monitoring.Notes:
+      type: object
+      properties:
+        notes_url:
```
Comment (kduret) [inline doc/API/centreon-api-v2.1.yaml:4242]:
```suggestion
        url:
```

File: doc/API/centreon-api-v2.1.yaml [hunk 14]
```diff
@@ -4232,11 +4234,21 @@ components:
...
+          type: string
+          nullable: true
+          description: "URL that can be used to provide more information on the resource"
+          example: "http://mediawiki/resource/name"
+        notes:
```
Comment (kduret) [inline doc/API/centreon-api-v2.1.yaml:4247]:
```suggestion
        label:
```

File: doc/API/centreon-api-v2.1.yaml [hunk 15]
```diff
@@ -581,6 +581,8 @@ paths:
         * status_code
         * status
         * action_url
+        * notes
```
Comment (kduret) [inline doc/API/centreon-api-v2.1.yaml:584]:
```suggestion
        * notes_label
```

File: doc/API/centreon-api-v2.1.yaml [hunk 16]
```diff
@@ -4232,11 +4234,21 @@ components:
...
-            notes_url:
-              type: string
-              nullable: true
-              description: "URL that can be used to provide more information on the resource"
-              example: "http://mediawiki/resource/name"
+            notes:
+              $ref: '#/components/schemas/Monitoring.Notes'
+    Monitoring.Notes:
+      type: object
+      properties:
```
Comment (kduret) [inline doc/API/centreon-api-v2.1.yaml:4241]:
```suggestion
      type: object
      nullable: true
      properties:
```

File: doc/API/centreon-api-v2.yaml [hunk 17]
```diff
@@ -4224,6 +4223,19 @@ components:
           type: string
           description: "Name of the group to which belongs the resource"
           example: "ESX-France|Cpus"
+    Monitoring.Notes:
+      type: object
+      properties:
```
Comment (kduret) [inline doc/API/centreon-api-v2.yaml:4228]:
```suggestion
      type: object
      nullable: true
      properties:
```

**Comment by adr-mo** [comment]
Force pushed done so that I can squash commits to ease the rebase on master (conflicts) with merge of https://github.com/centreon/centreon/pull/9516

File: doc/API/centreon-api-v2.yaml [hunk 18]
```diff
@@ -4229,6 +4228,19 @@ components:
...
```
Comment (kduret) [inline doc/API/centreon-api-v2.yaml:4233]:
```suggestion
      type: object
      nullable: true
      properties:
```

**Review by kduret — APPROVED** [review]

**Review by jeremyjaouen — APPROVED** [review]

**Pull request merged**